src/app-layer-ssl.c

```diff
@@ -287,11 +289,46 @@ static int TLSDecodeHandshakeHello(SSLState *ssl_state, uint8_t *input,
     if (!(HAS_SPACE(2)))
         goto invalid_length;
 
-    /* skip cipher suites */
-    uint16_t cipher_suites_length = input[0] << 8 | input[1];
-    input += 2;
+    if (ssl_state->curr_connp->handshake_type == SSLV3_HS_CLIENT_HELLO) {
+        uint16_t cipher_suites_length = input[0] << 8 | input[1];
+        input += 2;
+
+        if (cipher_suites_length & 1) {
+            SCLogDebug("Invalid length for cipher suites");
+            goto invalid_length;
+        }
+
+        if (!HAS_SPACE(cipher_suites_length)) {
+            SCLogDebug("Invalid length for cipher suites");
+            goto invalid_length;
+        }
+
+        ssl_state->curr_connp->num_cipher_suites = 0;
+        ssl_state->curr_connp->cipher_suites = NULL;
+
+        if (cipher_suites_length >= 2) {
+            ssl_state->curr_connp->num_cipher_suites = cipher_suites_length / 2;
+            ssl_state->curr_connp->cipher_suites = SCMalloc(cipher_suites_length);
+            if (!ssl_state->curr_connp->cipher_suites) {
+                SCLogError(SC_ERR_MEM_ALLOC, "Failed to allocate client cipher suites.");
+            } else {
+                memcpy(ssl_state->curr_connp->cipher_suites, input, cipher_suites_length);
+            }
+            input += cipher_suites_length;
+        }
+    }
+    if (ssl_state->curr_connp->handshake_type == SSLV3_HS_SERVER_HELLO) {
+        ssl_state->curr_connp->cipher_suites = SCMalloc(sizeof(uint16_t));
```

Comment:
I guess that we can check if cipher suites are already present and allocated in state struct.

"Double" handshake?